platforms/android/src/jp/wizcorp/phonegap/plugin/WizAssets/WizAssetsPlugin.java

```diff
@@ -98,114 +130,225 @@ public void run() {
-            }
+            deleteAssets(args, false, new DeleteAssetsCallback(callbackContext));
 
-            // We do not care about success or fail...
-            callbackContext.success();
             return true;
 
-        } else if (action.equals("deleteFile")) {
+        } else if (action.equals(DELETE_FILE_ACTION)) {
 
             Log.d(TAG, "[deleteFile] *********** " + args.getString(0));
-            String filepath = args.getString(0);
-
-            // If file is in bundle we cannot delete so ignore
-            if (!filepath.contains("www/assets")) {
-                // Delete the file, we do not care success or fail
-                // If you want to check file is deleted, delete() returns boolean
-                // so you can implement a check after delete() if necessary
-                File file = new File(filepath);
-                file.delete();
-                // Delete from database
-                wizAssetMan.deleteFile(filepath);
+            String filePath = args.getString(0);
+            try {
+                deleteAsset(filePath, false);
+            } catch (IOException e) {
+                callbackContext.error("Deleting file failed.");
+                return true;
             }
 
             // Callback success for any outcome.
             callbackContext.success();
             return true;
         }
+
         return false;  // Returning false results in a "MethodNotFound" error.
     }
 
-    @SuppressLint("InlinedApi")
-    private void downloadUrl(String fileUrl, String dirName, String fileName, String overwrite, CallbackContext callbackContext) {
+    private String getAssetFilePathFromUri(String file) {
+        String asset = wizAssetManager.getFile(file);
+        if (asset == null || asset == "" || asset.contains("NotFoundError")) {
+            return null;
+        }
+
+        return asset;
+    }
+
+    private String buildAssetFilePathFromUri(String uri) {
+        return pathToStorage + uri;
+    }
+
+    private void deleteAssets(JSONArray files, boolean isUri, DeleteAssetsCallback callback) {
+        AsyncDelete asyncDelete = new AsyncDelete(callback, isUri);
+        asyncDelete.execute(files);
+    }
+
+    private void deleteAsset(String filePath, boolean isUri) throws IOException {
+        // If file is in bundle we cannot delete so ignore and protect whole cache folder from being deleted
+        if (filePath != "" && !filePath.contains("www/assets")) {
+            if (isUri) {
+                filePath = buildAssetFilePathFromUri(filePath);
+            }
+            File file = new File(filePath);
+            deleteFile(file);
+            // Delete from database
+            wizAssetManager.deleteFile(filePath);
+        }
+    }
+
+    private void deleteFile(File file) throws IOException {
+        if(file.isDirectory()) {
+            //directory is empty, then delete it
+            if(file.list().length == 0){
```

Comment:
`file.list()` seems like an expensive directory traversal operation. You're doing it here, and on line 207. In the case of a non-empty folder, that means twice the cost. Actually three times, because you then repeat it on line 216. It seems to me that this function could be half the length and twice the speed.